org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Scanner.java

```diff
@@ -114,6 +115,16 @@ public class Scanner implements TerminalTokens {
 	public int[] lineEnds = new int[250];
 	public int linePtr = -1;
 	public boolean wasAcr = false;
+	// Positions in the source[] where \{ and } appear
+	public int[] embeddedExpressionPositions = new int[20];
+	// Contains a set of 4 positions
+	// 0: start pos of String fragment in this.withoutUnicodeBuffer, i.e. before /{
+	// 1: start pos of String fragment in this.source[], i.e. before /{
+	// 2: end pos of String fragment in this.withoutUnicodeBuffer i.e. after }
+	// 3: end pos of String fragment in this.source[], i.e. after }
+	public int[] fragmentPositions = new int[20];
+	public int embeddedExpressionPtr = 0;
+	public int fragmentPtr = 0;
 
```

Comment:
It ought to be sufficient to just maintain embeddedExpression start offset and end offset. If these offsets are relative to the string. At any point, we need to maintain these offsets only for the "current" "top level" string template. When we call parse expression on an embedded expression if that contains an embedded string template, that will become the "current" "top level" as we descend into it. Thus we will peel it layer by layer